# Intro to Data Science

[Gina Sprint](https://ginasprint.com/)

# Introduction to Machine Learning
What are our learning objectives for this lesson?
* Understand what machine learning is
* Revisit the concept of labeled and unlabeled data
* Understand the difference between supervised and unsupervised machine learning
* Understand the difference between classification and regression
* Learn about the kNN classification algorithm

Content used in this lesson is based upon information in the following sources:
* Dr. Shawn Bowers' Data Mining notes

## Warm up Task(s)
Open our MarkdownBasics.ipynb from last class: https://github.com/gsprint23/ZIME-Intro-to-Data-Science/blob/master/JupyterNotebookFun
1. Using Latex, typeset the quadratic formula:  
<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/c/c4/Quadratic_formula.svg/2560px-Quadratic_formula.svg.png" width="200"/>
1. Then, write a code cell with a function to implement it :)

## Today
1. A data storytelling with Jupyter Notebook example: https://github.com/gsprint23/ZIME-Intro-to-Data-Science/blob/master/JupyterNotebookFun/DataStorytellingExample.ipynb
1. Overview of k nearest neighbors algorithm
1. Break
1. Start ClassificationFun

## TODO
1. Work on Quiz 5 in Moodle
1. Work on project part 2 and project part 3 (BONUS)
    * Note: I moved project part 3 to be BONUS (extra credit) 😀

## Machine Learning
At a high level, machine learning is building and using models that are learned from data. Machine learning is a subset of artificial intelligence, and it greatly overlaps with data mining. Let's see the "unofficial" definitions for these areas from Wikipedia:
* [Data mining](https://en.wikipedia.org/wiki/Data_mining): The computational process of discovering patterns in large data sets involving methods at the intersection of artificial intelligence, machine learning, statistics, and database systems. It is an interdisciplinary subfield of computer science. The overall goal of the data mining process is to extract information from a data set and transform it into an understandable structure for further use.
    * Take away point: Discovering and using patterns in data
* [Artificial intelligence](https://en.wikipedia.org/wiki/Artificial_intelligence): The study of "intelligent agents": any device that perceives its environment and takes actions that maximize its chance of success at some goal. Colloquially, the term "artificial intelligence" is applied when a machine mimics "cognitive" functions that humans associate with other human minds, such as "learning" and "problem solving" (known as Machine Learning).
    * Take away point: Implementing human-cognition on a machine
* [Machine learning](https://en.wikipedia.org/wiki/Machine_learning): The subfield of computer science that, according to Arthur Samuel in 1959, gives "computers the ability to learn without being explicitly programmed." Evolved from the study of pattern recognition and computational learning theory in artificial intelligence, machine learning explores the study and construction of algorithms that can learn from and make predictions on data – such algorithms overcome following strictly static program instructions by making data driven predictions or decisions, through building a model from sample inputs.
    * Take away point: Learning from and making predictions on data
    
Here is an overview of the different components and applications of machine learning: 

![](https://miro.medium.com/max/1113/1*6iDmbHflsN6NULBLpVHA6A.png)

(image from https://miro.medium.com/max/1113/1*6iDmbHflsN6NULBLpVHA6A.png)

## Supervised Learning
Supervised learning requires "labeled" training data from a "supervisor." Such labels are considered the ground-truth for describing the data. The label comes from a knowledgeable expert and can be used to learn what information describes different labels.
* If the labeled attribute is categorical, then the learning task is called "classification"
* If the labeled attribute is continuous, then the learning task is called "regression"

Supervised learning is typically composed of training and testing. We will train a machine (AKA a student, learner, mathematical model) to learn a concept. Then we will test the machine's learned concept by applying their knowledge.

<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/0/09/Supervised_machine_learning_in_a_nutshell.svg/2000px-Supervised_machine_learning_in_a_nutshell.svg.png" width="650">

(image from [https://upload.wikimedia.org/wikipedia/commons/thumb/0/09/Supervised_machine_learning_in_a_nutshell.svg/2000px-Supervised_machine_learning_in_a_nutshell.svg.png](https://upload.wikimedia.org/wikipedia/commons/thumb/0/09/Supervised_machine_learning_in_a_nutshell.svg/2000px-Supervised_machine_learning_in_a_nutshell.svg.png))

### Training
As an example, suppose you are trying to teach someone (say a student) who has no notion of a cat or dog, the concept of cat vs. dog. You might first show the student some pictures of cats and say, "these are cats". Then you might show the person some pictures of dogs and say, "these are dogs". The set of cat and dog images is called the *training set*, a set of labeled examples (e.g. *instances*). For example, consider the following cat vs. dog training set:

<img src="https://raw.githubusercontent.com/gsprint23/cpts215/master/lessons/figures/cat_dog_training.png" width="500"/>

The student is going to look at different attributes of the image to try to learn a model of cat and a model of a dog. In doing so, the student will identify some aspects (AKA *attributes* or *features*) of the examples that distinguish a cat vs a dog. The features might include:

|Feature|Cat value|Dog value|
|-|-|-|
|Tongue out|No|Yes|
|Fur color|Light|Dark|
|Ears up|Yes|No|

What other features did you come up with?

#### Building a Model
A model to represent cat vs. dog based on these features might be rule-based:

>if tongue is out and the fur is dark and the ears are down then this is a dog

We will see later how we can use a tree with a rules (like the above) as a model to represent a classification such as dog vs. cat!

### Testing
Now, suppose we want to apply the student's learned conception of dog vs. cat by providing the student with a new, unseen example:

<img src="https://raw.githubusercontent.com/gsprint23/cpts215/master/lessons/figures/cat_or_dog.png" width="150"/>

Based on the above features, this image has the tongue out (dog), light fur color (cat), and ears up (cat). Thus our student would likely classify this image as a cat. But wait! We (the expert supervisors) know this is a dog (a puppy, but a dog none the less). Our training set didn't include any images that were as borderline cat/dog as this testing example. As you can see, the examples that comprise your training set and the features that are utilized greatly impact the accuracy of the learner, and consequently the model that is built. 

### Classification and Regression
The basic of idea of classification is:
* Given a data set (samples) and a new, unclassified instance
* Try to predict its classification (based on samples)

Note regression can be used in a similar way ... Let's say we have: $y = mx + b$

Q: How do we use this on a new instance?
* Predict a new $y'$ value from a new, unseen instance $x_{unseen}$ by calculating $y' = mx_{unseen} + b$

Approaches we will look at to classification
* k Nearest Neighbor (k-NN)... find "close cases"
* Naive Bayes... select "most probable" class for instance
* Decision Tree Induction... find "general" rules based on entropy
* Ensemble Methods... use many approaches to find best class (hybrid)

We'll also look at ways to evaluate classification results
* These largely involve splitting up a data set into training and testing sets
* Plus some basic statistics/metrics for accuracy, error

## Unsupervised Learning
Unsupervised learning does not require labeled training data. Information learned from the examples is data-driven and includes the process of discovering and describing patterns in the data. 

For example, to apply unsupervised learning to our cat vs. dog example, we would not try to "train" our student to learn the notion of "cat" or "dog". Instead, we would have our student look for patterns in the data, or perhaps a natural grouping. 

Here are our cat-dog training examples sorted in order based on the feature fur color:

<img src="https://raw.githubusercontent.com/gsprint23/cpts215/master/lessons/figures/cat_dog_fur_ordering.png" width="500"/>

We could apply a clustering algorithm, such as $k$-means clustering, to the data to reveal two natural groups in the data ($k = 2$):

<img src="https://raw.githubusercontent.com/gsprint23/cpts215/master/lessons/figures/cat_dog_grouping.png" width="500"/>

Note that these two groups, blue and red, are not representative of cat and dog, since we have no cat/dog labels!

Now, upon seeing a new instance, we can determine the new instance's membership to either the blue group or the red group:

<img src="https://raw.githubusercontent.com/gsprint23/cpts215/master/lessons/figures/cat_dog_membership.png" width="500"/>

Like supervised machine learning, there are several unsupervised machine learning algorithms.

## Nearest Neighbor Classification
Nearest neighbor classification is typically used when attributes are continuous
* Can be modified for categorical data ...

Basic approach:
* Given an instance $i$ with $n - 1$ attributes (where the $n^{th}$ is class label)
* Find the "closest" instance $j$ to $i$ on the $n - 1$ attributes
* Use $j$'s class as the prediction for $i$

Example from [Bramer](https://www.amazon.com/Principles-Mining-Undergraduate-Computer-Science/dp/1447148835):
Given the data set

|a |b |c |d |e |f |Class|
|-|-|-|-|-|-|-|
|yes |no |no |6.4 |8.3 |low |negative|
|yes |yes |yes |18.2 |4.7 |high |positive|

What should this instance's classification be?

|yes |no |no |6.6 |8.0 |low |???|
|-|-|-|-|-|-|-|

Usually it isn't this easy!

## k Nearest Neighbors
Find the k nearest neighbors ...
* Usually find the k closest neighbors (instead of just closest)
* Then pick classification from among the top k

What are good values for the number of neighbors k?
* Often done experimentally (with a test set)
* Start with k = 1, determine "error rate" (more later)
* Repeat incrementing k
* Pick k with smallest (minimum) error rate
     * Often, larger the data (training) set, the larger the k

Lets say we found the k nearest neighbors for an instance ...

Q: What are ways we could pick the class?
* Most frequent occurring class
* Weighted "average" (based on the relative closest of the k)

Note: can use k-NN for regression if, e.g., return the mean of the label values

## Distance Functions
k-NN works by calculating distances between instances
* Many possible ways to do this ... generalized through "distance measures"
* For two points $x$ and $y$, the distance between them is given by $dist(x,y)$

Properties of distance measures (metrics)
1. $\forall x$, $dist(x,x) = 0$
    * The distance of any point x from itself is zero
2. $\forall xy$, $dist(x,y) = dist(y,x)$
    * Symmetry
3. $\forall xyz$, $dist(x,y) \leq dist(x,z) + dist(z,y)$
    * Triangle equality
    * "Shortest distance between any two points is a straight line"

Euclidean Distance is most often used: Given an instance, treat it as a "vector" in n space
* Use Pythagoras' Theorem to find distance between them

For example:
* Given two points (i.e., rows) with $n = 2$: $(x_1, y_1)$ and $(x_2, y_2)$
* The length (i.e., distance) of the straight line joining the points is:

$$\sqrt{(x_1 - x_2)^{2} + (y_1 - y_2)^{2}}$$

* Euclidean $n$-space
    * For rows A = $(a_1, a_2,..., a_n)$ and $B = (b_1, b_2,..., b_n)$ with $n$ attributes
$$\sqrt{(a_1 - b_1)^{2} + (a_2 - b_2)^{2} +...+ (a_n - b_n)^{2}}$$
        * Which is:
$$\sqrt{\sum_{i=1}^{n}(a_i - b_i)^{2}}$$

Other examples are described in the book (e.g., Manhattan "city block" distance)

Q: Do you see any possible issues with Euclidean distance?
* Larger values tend to dominate smaller ones
* Can degrade into a few attributes driving the distances
    * e.g., [Mileage=18,457, Doors=2, Age=12]
    
## Normalization
One solution is to scale all values between 0 and 1 ("min-max" normalization)
* Use the formula: `(x - min(xs)) / ((max(xs) - min(xs)) * 1.0)`

Q: How can we deal with categorical values?
* $dist(v_1, v_2) = 0$ if values $v_1 = v_2$
    * Same values
* For nominal values, $dist(v_1, v_2) = 1$ if $v_1 \neq v_2$
    * Different values
* For ordinal values, assign to 1 or use the "distance"

Q: What do we do about missing values?
* Don't have missing values
    * Clean the data first
* Be conservative
    * Assuming normalized
    * If only one value missing:
        * Assume the maximum possible distance
        * If nominal use the maximum distance (i.e., 1)
        * If ordinal use either 1 or furthest distance from known value
    * otherwise, if both values missing, use the maximum distance (e.g., 1)

Distance-based metrics imply equal weighting of attributes
* Sometimes can perform better with attribute weights (i.e., some attributes worth more)
* "Feature reduction" (not using certain attributes) can also help with redundant or "noisy" attributes

## Basic k-NN Algorithm
```
Input: list of rows, no of atts (n where nth is label), instance to classify, k
def kNN_classifier(training_set, n, instance, k):
    row_distances = []
    for row in training_set:
        d = distance(row, instance, n - 1)
        row_distances.append([d, row])
    top_k_rows = get_top_k(row_distances, k)
    label = select_class_label(top_k_rows)
    return label
```

## kNN Example
Example adapted from [this kNN example](https://people.revoledu.com/kardi/tutorial/KNN/KNN_Numerical-example.html)

Suppose we have the following dataset that has two attributes (acid durability and strength) and a class attribute (whether a special paper tissue is good or not):

|Acid durability (seconds)|Strength (kg/square meter)|Classification|
|-|-|-|
|7|7|Bad|
|7|4|Bad|
|3|4|Good|
|1|4| Good|

Now the factory produces a new paper tissue with acid durability = 3 seconds and strength = 7 kg/square meter. Can we predict what the classification of this new tissue is? Use kNN with $k$ = 3. 

### Make a Prediction Manually
Steps:
1. Normalize
1. Compute distance of each training instance to the test instance
1. Determine the majority classification of the $k$ closest instances... this is your prediction for the test instance

After normalization:

|Acid durability (seconds)|Strength (kg/square meter)|Classification|
|-|-|-|
|1|1|Bad|
|1|0|Bad|
|0.33|0|Good|
|0|0| Good|

Test instance normalization: 0.33, 1

Distances:

|Acid durability (seconds)|Strength (kg/square meter)|Classification|Distance|
|-|-|-|-|
|1|1|Bad|0.66|
|1|0|Bad|1.203|
|0.33|0|Good|1.0|
|0|0| Good|1.05|

Work:
* $\sqrt{(1-0.33)^2 + (1-1)^2} = 0.66$
* $\sqrt{(1-0.33)^2 + (0-1)^2} = 1.203$
* $\sqrt{(0.33-0.33)^2 + (0-1)^2} = 1.0$
* $\sqrt{(0-0.33)^2 + (0-1)^2} = 1.05$

Majority classification: 
1 Bad (0.66) and 2 Goods (1.0 an 1.05) => Good!

### Make a Prediction with Scikit-Learn
Steps:
1. Load data
1. Normalize
1. Train kNN classifier with training set
1. Test kNN classifier on test instance

In [1]:
# load data
import pandas as pd

data = [[7, 7, "Bad"], [7, 4, "Bad"], [3, 4, "Good"], [1, 4, "Good"]]
df = pd.DataFrame(data, columns=["Acid durability (seconds)", "Strength (kg/square meter)", "Classification"])

print(df)

   Acid durability (seconds)  Strength (kg/square meter) Classification
0                          7                           7            Bad
1                          7                           4            Bad
2                          3                           4           Good
3                          1                           4           Good


In [7]:
# normalize
from sklearn.preprocessing import MinMaxScaler

X_train = df.drop("Classification", axis=1)
y_train = df["Classification"]

scaler = MinMaxScaler()
scaler.fit(X_train)
print(scaler.data_min_)
print(scaler.data_max_)

X_train_normalized = scaler.transform(X_train)
print(X_train_normalized)

[1. 4.]
[7. 7.]
[[1.         1.        ]
 [1.         0.        ]
 [0.33333333 0.        ]
 [0.         0.        ]]


In [8]:
# train
from sklearn.neighbors import KNeighborsClassifier
neigh = KNeighborsClassifier(n_neighbors=3)
neigh.fit(X_train_normalized, y_train)

# test
X_test = pd.Series([3, 7], index=df.columns.drop("Classification"))
X_test = scaler.transform([X_test])
y_test_prediction = neigh.predict(X_test)
print(y_test_prediction)

['Good']


### Some Notes
Q: What happens if there are ties in the top-k distances (get_top_k)? E.g., which are top 3 in: [[.28,$r_1$],[.33,$r_2$],[.33,$r_3$],[.33,$r_4$],[.37,$r_5$]]?
* Different options ... e.g.:
    * Randomly select from ties
    * Do top-k distances (instead of instances)
    * Ignore ties (in case above, just use $r_1$ and $r_2$)

Nearest doesn't imply near
* top-k instances might not be that close to the instance being classified
* Especially true as the number of attributes ("dimensions") increases
    * An example of the "curse of dimensionality"
* Again, have to use common sense and an understanding of the dataset

### Efficiency issues
Q: Is k-NN efficient? Can you find any efficiency issues?
* Given a training set with $D$ instances and $k = 1$
* $O(D)$ comparisons needed to classify a given instance

Q: Can you think of any ways to improve the efficiency?
1. Use search trees
    * Presort and arrange instances into a search tree
    * Can reduce comparisons to $O(log D)$
2. Check each training instance in parallel
    * Gives $O(1)$ comparisons
3. Editing/Pruning
    * Filter or remove training tuples that prove useless
    * Reduces size of $D$